local.go

```diff
@@ -275,8 +275,17 @@ func checkConvergence(loc *Location, itertype IterationType, stats *Stats, setti
 	return NotTerminated
 }
 
-// evaluate evaluates the function and stores the answer in place
+// evaluate evaluates the function given by funcs at xNext, stores the answer
+// in location and updates stats. If the function does not support the
+// requested evalType, location and stats are not modified and ErrMismatch is
+// returned.
 func evaluate(funcs functions, funcInfo *FunctionInfo, evalType EvaluationType, xNext []float64, location *Location, stats *Stats) error {
+	if evalType == NoEvaluation {
+		return nil
+	}
+	if !funcInfo.Supports(evalType) {
```

Comment:
> I think this should panic.

Agreed. I just used the error that had already been there.

> For example, if the function is constrained and the user chooses Lbfgs, there should be an err/panic.

If the function is also a Function, then it does not necessarily need to err/panic in that case. This is something to keep for later.

> In the end, it's not about Function supporting EvalType, but Method supporting the specific Function.

Agreed. I will remove Supports() from FunctionInfo.

> Given we do use some magic, I can see the need for support. It seems like the better case is for the optimizer to have a Support function that either returns a boolean (if it supports), or an error which is nil if it supports, and provides a detailed explanation if it isn't, i.e. "Bfgs requires the Function to have a Gradient. Function does not implement either Gradient or FunctionGradient."

Good idea, I will put it into Issues as an enhancement.